File: ruby/recursed/recur.rb
#get user input
print "Enter the upper bound: "
n = STDIN.gets.chomp.to_i

#initialize array to hold the largest
#used sorted tuples to store values
largest = [[1, 0]]

#function for getting the sequence length
#recursive version
def collatz(n)
    if(n == 1)
        return 0
    else
        if(n % 2 == 0)
            return collatz(n/2) + 1
        else
            return collatz(3*n + 1) + 1
        end
    end
end

#main loop
for i in 1..n do
    #get the current sequence length
    l = collatz(i)
    #get the least significant sequence length of stored values
    temp = largest[0][1]

    #if current is more significant
    if l > temp
        if largest.length >= 10
            #replace if array is holding ten elements
            largest.shift
        end
        #otherwise add to end
        e = [i, l]
        largest.push(e)

        #sort by value descending, then by sequence length ascending
        largest = largest.sort { |a, b| [a[1], a[0]] <=> [b[1], b[0]] }
        #result is the least significant element at index 0
    end
end

#sort by value ascending, then by sequence length descending
largest = largest.sort { |a, b| [b[1], a[0]] <=> [a[1], b[0]] }

#print the stored values
for e in largest do
    puts "#{e[0]} #{e[1]}"
end

File: python/recursed/recur.py
#get user input
n = int(input("Enter the upper bound: "))

#initialize array to hold the largest
#used sorted tuples to store values
largest = [(1, 1)]

#function for getting the sequence length
#recursive version
def collatz(num):
    if(num == 1):
        return 0
    else:
        if(num % 2 == 0):
            return collatz(num/2) + 1
        else:
            return collatz(3*num + 1) + 1

#main loop
for i in range(1, n+1):
    #get the current sequence length
    length = collatz(i)
    #get the least significant sequence length of stored values
    temp = largest[-1]

    #if current is more significant
    if(length > temp[1]):
        if(len(largest) >= 10):
            #replace if array is holding ten elements
            largest.remove(temp)
        #otherwise add to end
        largest.append((i, length))

        #sort by value ascending, then by sequence length descending
        largest.sort(key=lambda l:l[0])
        largest.sort(key=lambda l:l[1], reverse = True)
        #result is the least significant element at the last index

#print the stored values
for i in range(len(largest)):
    print(largest[i][0], largest[i][1])
